src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
+        }
+        else
+        {
+            pContext->MultiSz.Append(*pstruConfigPath);
+        }
+    }
+    return fChanged;
+}
+
+//
+// Calls recycle on the application in the application manager
+// Will 
 HRESULT
-APPLICATION_MANAGER::RecycleApplication(
+APPLICATION_MANAGER::RecycleApplicationFromManager(
     _In_ LPCWSTR pszApplicationId
 )
 {
     HRESULT          hr = S_OK;
     APPLICATION_INFO_KEY  key;
     DWORD            dwPreviousCounter = 0;
+    APPLICATION_INFO_HASH* table = NULL;
+    CONFIG_CHANGE_CONTEXT context;
 
     hr = key.Initialize(pszApplicationId);
     if (FAILED(hr))
     {
         goto Finished;
     }
+
+    table = new APPLICATION_INFO_HASH();
+
+    if (table == NULL)
+    {
+        hr = E_OUTOFMEMORY;
+        goto Finished;
+    }
+
+    //
+    // few application expected, small bucket size for hash table
+    //
+    if (FAILED(hr = table->Initialize(17 /*prime*/)))
+    {
+        goto Finished;
+    }
+
+    context.pstrPath = pszApplicationId;
+
     AcquireSRWLockExclusive(&m_srwLock);
     dwPreviousCounter = m_pApplicationInfoHash->Count();
 
-    m_pApplicationInfoHash->DeleteKey(&key);
+    // We don't want to hold the lock for long time as it will block all incoming requests
+    // Make a shallow copy of existing hashtable as we may nee to remove nodes
+    // This also make sure application shutdown will not be called inside the lock
+    m_pApplicationInfoHash->Apply(APPLICATION_INFO_HASH::ReferenceCopyToTable, static_cast<PVOID>(table));
+    DBG_ASSERT(dwPreviousCounter == table->Count());
+    
+    // Removed the applications which are impacted by the configurtion change
+    m_pApplicationInfoHash->DeleteIf(FindConfigChangedApplication, (PVOID)&context);
 
-    if(dwPreviousCounter != m_pApplicationInfoHash->Count())
+    if (dwPreviousCounter != m_pApplicationInfoHash->Count())
     {
-        // Application got recycled. Log an event
-        STACK_STRU(strEventMsg, 256);
-        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-            ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
-            pszApplicationId)))
+        if (m_hostingModel == HOSTING_IN_PROCESS)
         {
-            UTILITY::LogEvent(g_hEventLog,
-                EVENTLOG_INFORMATION_TYPE,
-                ASPNETCORE_EVENT_RECYCLE_CONFIGURATION,
-                strEventMsg.QueryStr());
+            // When we are inprocess, we need to keep the application the
+            // application manager that is being deleted. This is because we will always need to recycle the worker
+            // process and any requests that hit this worker process must be rejected (while out of process can
+            // start a new dotnet process). We will immediately call Recycle after this call.
+
+            // We need to clear the hash table here and the ref copy the original table to avoid memory leaks
+            // and keep the application referenced.
+            m_pApplicationInfoHash->Clear();
```

Comment:
no need to call clear. For inprocess, m_pApplicationInfoHash should contain zero entry. You can do DBG_ASSERT(m_pApplicationInfoHash->Count() ==0)